Tests/NIOTests/EventLoopFutureTest.swift

```diff
@@ -828,4 +828,65 @@ class EventLoopFutureTest : XCTestCase {
         XCTAssertTrue(noHoppingFuture === noHoppingPromise.futureResult)
         noHoppingPromise.succeed(result: ())
     }
+
+    func testWhenAllCompleteResultsWithFailuresStillSucceed() {
+        let group = MultiThreadedEventLoopGroup(numberOfThreads: 2)
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+
+        let future: EventLoopFuture<[Result<Bool, Error>]> = .whenAllComplete([
+            group.next().makeFailedFuture(error: EventLoopFutureTestError.example),
+            group.next().makeSucceededFuture(result: true)
+        ], eventLoop: group.next())
+        XCTAssertNoThrow(try future.wait())
+    }
+
+    func testWhenAllCompleteResults() throws {
+        let group = MultiThreadedEventLoopGroup(numberOfThreads: 2)
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+
+        let results = try EventLoopFuture<[Result<Int, Error>]>.whenAllComplete([
+            group.next().makeSucceededFuture(result: 3),
+            group.next().makeFailedFuture(error: EventLoopFutureTestError.example),
+            group.next().makeSucceededFuture(result: 10),
+            group.next().makeFailedFuture(error: EventLoopFutureTestError.example),
+            group.next().makeSucceededFuture(result: 5)
+        ], eventLoop: group.next()).wait()
+
+        XCTAssertEqual(try results[0].get(), 3)
+        XCTAssertThrowsError(try results[1].get())
+        XCTAssertEqual(try results[2].get(), 10)
+        XCTAssertThrowsError(try results[3].get())
+        XCTAssertEqual(try results[4].get(), 5)
+    }
+
+    func testWhenAllCompleteResolvesAfterFutures() throws {
+        let group = MultiThreadedEventLoopGroup(numberOfThreads: 6)
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+
+        var tally = 0
+        
+        var futures: [EventLoopFuture<Int>] = []
+        for index in 1...5 {
+            futures.append(
+                group.next().makeSucceededFuture(result: index).map {
+                    sleep(UInt32(1 / index))
```

Comment:
There's no need to `sleep` here: all that does is slow down the test unnecessarily.

The better approach is to do something like this:

```swift
let promises = (0..<5).map { _ in group.next().makePromise(of: Int.self) }
let futures = promises.map { $0.futureResult }

var succeeded = false
var completedPromises = false

let mainFuture = EventLoopFuture.whenAllComplete(futures, eventLoop: group.next())
mainFuture.whenSuccess { result in
    XCTAssertTrue(completedPromises)
    XCTAssertFalse(succeeded)
    succeeded = true
}

XCTAssertFalse(succeeded)

// Complete first four promises.
for (index, promise) in promises.dropLast().enumerated() {
    promise.succeed(result: idx)
}

// Still hasn't succeeded.
XCTAssertFalse(succeeded)

// Complete last promise.
completedPromises = true
promises.last!.succeed(result: 4)

let result = try assertNoThrowWithValue(mainFuture.wait())
XCTAssertEqual(result, [0, 1, 2, 3, 4])
```

This test is much more deterministic, observes all the intermediate states, and doesn't need to sleep unnecessarily.